### PR TITLE
Wip/hughsie/fu firmware add patch

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: libfwupdplugin/fu-firmware.c [hunk 1]
```diff
@@ -1038,13 +1099,52 @@ fu_firmware_write(FuFirmware *self, GError **error)
...
+
+	/* ensure exists */
+	if (priv->patches == NULL)
+		priv->patches =
+		    g_ptr_array_new_with_free_func((GDestroyNotify)fu_firmware_patch_free);
+
+	/* find existing */
+	for (guint i = 0; i < priv->patches->len; i++) {
+		ptch = g_ptr_array_index(priv->patches, i);
+		if (ptch->offset == offset) {
```
Comment (superm1) [inline libfwupdplugin/fu-firmware.c:1136]: You need to make sure the new patch is the same length or you might be throwing away part of the old one.

Comment (hughsie) [reply libfwupdplugin/fu-firmware.c:1136]: Absolutely correct, fixed. Do you think we need a `fu_firmware_remove_patch` (just taking the offset) too?

Comment (superm1) [reply libfwupdplugin/fu-firmware.c]: I don't know how important that is - users can just re-run to redo patches right?

Comment (hughsie) [reply libfwupdplugin/fu-firmware.c]: Right, just from a programmatic API point of view.

Comment (superm1) [reply libfwupdplugin/fu-firmware.c]: I don't see harm in it if you think something might use it some day.

The only use case I can think of is some vendor wants to ship patches inside a cab and the patches are applied by the software. If that's not a good idea then I would say no need.

Comment (hughsie) [reply libfwupdplugin/fu-firmware.c]: > The only use case I can think of

Ya, I couldn't think of a use case either; lets add it if/when we need it. Lets see if this works for the reporter.

**Review by superm1 — DISMISSED** [review]
LGTM if this meets the needs from OP

**Pull request merged**